### PR TITLE
Fix WebPack 2 query object in loader chain

### PR DESCRIPTION
**Pull request by texastoland**
- Support WebPack 2 query objects.
- Migrate APIs to accept single options object instead of variably ordered args per WebPack 2 convention. Necessary to differentiate between loader and options objects.
- Copy error message from `CommonChunksPlugin`.
- Delegate `extract` class function to instance method to eliminate copy pasta. Note that `loader` delegates the opposite direction.
- Add query object test using CSS source maps.
- Port original example to WebPack 2.
- Update README

Will publish if necessary CC @andreypopp @antialias @gajus @haoxins @indeyets @jamiehodge @jsdf @ju1ius @kentcdodds @spicydonuts @timbru31

### CODE REVIEW

File: loader.js [hunk 1]
```diff
@@ -36,89 +36,84 @@ module.exports.pitch = function(request) {
 			resultSource = undefined;
 		}
 
-		if(query.extract !== false) {
```
Comment (AppShipIt) [inline loader.js]: Nixed this line. The rest of this file is just decreased indentation.

**Comment by AppShipIt** [comment]
Not much I can do to improve patch coverage. The vast majority of missed lines are handling errors.

File: index.js [hunk 2]
```diff
@@ -100,16 +100,28 @@ function getOrder(a, b) {
...
 }
 
-function ExtractTextPlugin(id, filename, options) {
-	if(typeof filename !== "string") {
-		options = filename;
-		filename = id;
-		id = ++nextId;
+function ExtractTextPlugin(options) {
+	if(arguments.length > 1) {
+		throw new Error("Deprecation notice: ExtractTextPlugin now only takes a single argument. Either an options " +
```
Comment (kentcdodds) [inline index.js]: `Deprecation`s normally mean that: "This still works the way you're using it, but it'll stop working soon." Whereas this PR causes the old API to not work at all. I recommend either 1) we make it support both APIs, 2) change the words `Deprecation notice` to `Breaking change`

Comment (kentcdodds) [inline index.js]: I should also say, this is a great message. Good work :+1:

File: index.js [hunk 3]
```diff
@@ -154,28 +153,52 @@ ExtractTextPlugin.prototype.applyAdditionalInformation = function(source, info)
...
-		].concat(before, loader).join("!");
-	} else {
-		options = loader;
-		loader = before;
-		return [
-			this.loader(mergeOptions({remove: true}, options))
-		].concat(loader).join("!");
+ExtractTextPlugin.prototype.extractAll = function(options) {
+	if(arguments.length > 1) {
+		throw new Error("Deprecation notice: extract now only takes a single argument. Either an options " +
```
Comment (kentcdodds) [inline index.js]: Same thing here as above

File: loader.js (hunk 1 above)
Comment (kentcdodds) [inline loader.js]: Thanks for saying that. This diff was intimidating :-)

File: package.json [hunk 4]
```diff
@@ -4,7 +4,7 @@
   "author": "Tobias Koppers @sokra",
   "description": "Extract text from bundle into a file.",
   "peerDependencies": {
-    "webpack": "^1.9.11"
+    "webpack": "^1.9.11 || ^2.1.0-beta"
```
Comment (kentcdodds) [inline package.json]: So this technically still works with webpack 1? That's great 👍

**Comment by kentcdodds** [comment]
I don't really know the codebase well, but I don't see anything wrong with the PR

File: package.json (hunk 4 above)
Comment (AppShipIt) [inline package.json]: All tests pass on both 🎉

File: loader.js (hunk 1 above)
Comment (AppShipIt) [inline loader.js]: Lesson learned next time I'll split it into two commits (delete and reformat) so git doesn't get confused.

File: index.js (hunk 3 above)
Comment (AppShipIt) [inline index.js]: Good point about <q>breaking change</q> I should add a migration note to the README as well.

**Comment by AppShipIt** [comment]
API discussion at https://gitter.im/webpack/webpack/developer?at=5763aed963ea0987306be32d.

File: index.js (hunk 2 above)
Comment (ianks) [inline index.js]: Hehe thanks @AppShipIt. I'm happy to see webpack getting better error messages!

**Comment by AppShipIt** [comment]
It's tested in our app but not deployed yet @sokra 🙏

**Comment by robwise** [comment]
I just went through @kentcdodds egghead tutorial using webpack 2, but we can't deploy to production because I can't shrinkwrap without the peer dependency issue resolved. This would be really awesome to get this merged!

**Comment by AppShipIt** [comment]
We'll publish our fork soon if we don't hear back. My impression is @sokra doesn't merge big PRs.

File: index.js [hunk 5]
```diff
@@ -154,27 +160,45 @@ ExtractTextPlugin.prototype.applyAdditionalInformation = function(source, info)
...
-		loader = before;
-		return [
-			this.loader(mergeOptions({remove: true}, options))
-		].concat(loader).join("!");
+ExtractTextPlugin.prototype.extract = function(options) {
+	if(arguments.length > 1) {
+		throw new Error("Breaking change: extract now only takes a single argument. Either an options " +
+						"object *or* the loader(s).\n" +
+						"Example: if your old code looked like this:\n" +
+						"    ExtractTextPlugin.extract('syle-loader', 'css-loader')\n\n" +
```
Comment (s-panferov) [inline index.js]: `s/syle/style/`

Comment (TheLarkInn) [inline index.js]: Nice catch!

Comment (texastoland) [inline index.js]: Rebasing 🙏 Still waiting on @sokra but planning to publish Friday otherwise.

**Comment by texastoland** [comment]
To anyone subscribed I'll be deleting my fork next week consequently shutting this PR. WebPack's maintainer has not only been unresponsive on multiple communication channels but tagged a new beta without addressing any outstanding PR notably including this one for WebPack 2 migration. Besides losing confidence in its maintainer our build broke irrevocably from upstream dependencies and lack of communication nor accepting meaningful contributions as evident in each repo's activity graph. Above all having delved into the source on several necessary occasions it's a bowl of spaghetti I won't bank our team's productivity on. He doesn't owe anyone support but I cannot commend his software either.

**Comment by TheLarkInn** [comment]
@texastoland before you jump ship, please reconsider reopening.

**Comment by kentcdodds** [comment]
> @texastoland before you jump ship, please reconsider reopening.

I appreciate what you're trying to do @TheLarkInn, but @texastoland is totally right in his decision. This has been a bit of a fiasco. I hope that the webpack project at large can improve the way it works with the community so this kind of thing doesn't happen again.

**Comment by TheLarkInn** [comment]
So I'm to blame for lack of communication. First and foremost getting that out of the way. I apoligize. 

@kentcdodds @texastoland here's the low down.
We have discussed this repo for 2 weeks now. ExtractTextPlugin is a pain point for many people, whether it be speed, hackyness, bugs, etc. We have been so busy with documentation efforts and some of us being away or busy with real-life that it would be harmful to not spend the time looking over a PR in detail, especially one that already has problems. Communication certainly could have happened. I will personally ensure that this will certainly not happen again. Fuethermore in the future once we have assigned all new maintainers to these repositories, we can have a bit more leeway, flexibility, and autonomy. And if anyone has any trouble with this can personally reach out to me.

**Pull request merged**